src/sem_hon_psi.rs

```diff
@@ -0,0 +1,147 @@
+/// This crate implements the semi-honest private set intersection (PSI) protocol described in
+/// page 14 on the PDF at https://cyber.biu.ac.il/wp-content/uploads/2017/01/15.pdf
+///
+/// There are two parties, the sender and the receiver, and each of them have a set of integers (WLOG).
+/// This PSI protocol reveals the intersection of these sets to the receiver.
+/// 
+/// The protocol starts by the receiver running `receiver_1` and sending the output to
+/// the sender. Then, the sender uses the data it receives as input to `sender_1`.
+/// Finally, the receiver runs `receiver_2` with the output of `sender_1`
+/// and the output of `receiver_2` shows the intersection to the receiver.
+
+use ark_ff::{
+    fields::{Field, Fp64, MontBackend, MontConfig},
+    BigInteger256, PrimeField
+};
+
+use rand::{thread_rng, Rng, rngs::ThreadRng};
+
+// https://docs.rs/sha2/latest/sha2/
+use sha2::{Sha256, Digest};
+
+// Defining your own field
+// To demonstrate the various field operations, we can first define a prime ordered field $\mathbb{F}_{p}$ with $p = 17$. When defining a field $\mathbb{F}_p$, we need to provide the modulus(the $p$ in $\mathbb{F}_p$) and a generator. Recall that a generator $g \in \mathbb{F}_p$ is a field element whose powers comprise the entire field: $\mathbb{F}_p =\\{g, g^1, \ldots, g^{p-1}\\}$.
+// We can then manually construct the field element associated with an integer with `Fp::from` and perform field addition, subtraction, multiplication, and inversion on it.
+#[derive(MontConfig)]
+#[modulus = "2305843009213693951"] // a Mersenne prime
+#[generator = "3"]
+pub struct FqConfig;
+pub type Fq = Fp64<MontBackend<FqConfig, 1>>;
+
+
+/// Called by the receiver to start the protocol.
+/// 
+/// # Arguments
+///
+/// * `set_y` - The receiver's private set
+/// 
+/// # Outputs
+///
+/// * `alpha` - Random integer. Will be later used in `receiver_2`
+/// * `set_y_alpha` - Set generated by raising each element in `set_y` by `alpha`
+pub fn receiver_1(set_y: &Vec<u64>) -> (Vec<Fq>, BigInteger256) {
+    let mut rng: ThreadRng = thread_rng();
+
+    // receiver's randomness
+    let alpha: BigInteger256 = rng.gen();
+
+    // NOTE: this entire code will be part of a zk circuit, so we avoid optimizations
+    let mut set_y_alpha: Vec<Fq> = Vec::new();
+    for elt in set_y {
```

Comment:
could do this with a `.map()` to be more functional (read rust-y)

Reply:
Yes, but I will wrap everything in ZK, so the `.map()` would probably not be transferable. So, didn't work on it.